### PR TITLE
hooks: add hook for rtree (rokm)

### PR DESCRIPTION
**Pull request by eric15342335**
https://github.com/pyinstaller/pyinstaller/discussions/6060#discussioncomment-1142925

### CODE REVIEW

**Comment by eric15342335** [comment]
the workflow isn't passing.

File: src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# This file is distributed under the terms of the GNU General Public
+# License (version 2.0 or later).
+#
+# The full license is available in LICENSE.GPL.txt, distributed with
+# this software.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+# ------------------------------------------------------------------
+
+from PyInstaller.utils.hooks import is_pure_conda
```
Comment (bwoodsend) [inline src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py]: This should be:
```suggestion
from PyInstaller.compat import is_pure_conda
```

File: src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# This file is distributed under the terms of the GNU General Public
+# License (version 2.0 or later).
+#
+# The full license is available in LICENSE.GPL.txt, distributed with
+# this software.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+# ------------------------------------------------------------------
+
+from PyInstaller.compact import is_pure_conda
```
Comment (bwoodsend) [inline src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py:13]:
```suggestion
from PyInstaller.compat import is_pure_conda
```

**Review by bwoodsend — CHANGES_REQUESTED** [review]

**Comment by bwoodsend** [comment]
I notice the docs for `conda_support` are wrong in that they suggest to do what you did and import `is_pure_conda` from `PyInstaller.utils.hooks`.

**Comment by eric15342335** [comment]
Yea there's what I came from

**Review by rokm — CHANGES_REQUESTED** [review]
Hmm, I think this should be reworked a bit...

First, we probably want to call `PyInstaller.utils.hooks.collect_dynamic_libs('rtree')` in all cases, because the user might have installed an `rtree` wheel from PyPI (for whatever reason) within their conda environment (and if I recall correctly, `is_pure_conda` won't catch that, because it distinguishes only between pure conda env and virtualenv within conda env).

Second, have you checked the output of `binaries = PyInstaller.utils.hooks.conda.collect_dynamic_libs('rtree', dest='rtree/lib')` ? That's going to be ugly and will add unnecessary increase the size to the bundled program. 

I suppose the entries need to be filtered, so that only `spatialindex*.dll` files are collected. 

Alternatively, we could skip collect_dynamic_libs and try to glob for `os.path.join(sys.prefix, 'Library', 'bin', 'spatialindex*.dll')`.

Or perhaps the best approach would be to instead use `binaries = PyInstaller.utils.hooks.conda.collect_dynamic_libs('libspatialindex', dest='rtree/lib', dependencies=False)`

**Comment by eric15342335** [comment]
I made the history worser

**Comment by bwoodsend** [comment]
> Or perhaps the best approach would be to instead use binaries = PyInstaller.utils.hooks.conda.collect_dynamic_libs('libspatialindex', dest='rtree/lib', dependencies=False)

I imagine this would be best. I'm having second thoughts about that default of `dependencies=True`.

**Comment by bwoodsend** [comment]
> Yea that's where I came from, maybe a docs fix?

Yes please!

**Comment by rokm** [comment]
Also, we should probably enter the conda-specific lookup only if `binaries` is still empty after the preceding call to `PyInstaller.utils.hooks.collect_dynamic_libs('rtree')`.

File: src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py [hunk 3]
```diff
@@ -11,10 +11,8 @@
...
+from PyInstaller.utils.hooks import collect_dynamic_libs, conda
 
-if is_pure_conda:
-    from PyInstaller.utils.hooks.conda import collect_dynamic_libs
-    binaries = collect_dynamic_libs('rtree', dest='rtree/lib')
-else:
-    from PyInstaller.utils.hooks import collect_dynamic_libs
-    binaries = collect_dynamic_libs('rtree', destdir='rtree/lib')
+binaries = collect_dynamic_libs('rtree', destdir='rtree/lib')
+if not binaries and is_pure_conda:
```
Comment (rokm) [inline src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py:17]: Is there a reason for using `is_pure_conda` instead of just `is_conda`?

Comment (bwoodsend) [reply src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py]: That should be `is_pure_conda`. The `conda_support` module doesn't exist if the user is running venv inside a conda environment.

Comment (eric15342335) [reply src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py]: I thought probably this
https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/compat.py#L165

Comment (bwoodsend) [reply src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py]: Yeah, `rtree` can't be a conda distribution unless `is_pure_conda` is true.

File: src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# ------------------------------------------------------------------
+# Copyright (c) 2020 PyInstaller Development Team.
```
Comment (rokm) [inline src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py:2]: Final nitpick: it's 2021 (even though these days time might appear to have frozen).

Comment (eric15342335) [reply src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-rtree.py:2]: holy i copied the wrong one

**Review by rokm — APPROVED** [review]

**Pull request merged**